cpp/src/arrow/dataset/scanner.cc

```diff
@@ -151,20 +154,79 @@ Status ScannerBuilder::Finish(std::unique_ptr<Scanner>* out) const {
   return Status::OK();
 }
 
-Status Scanner::ToTable(std::shared_ptr<Table>* out) {
+static inline Status SerialToTable(ScanTaskIterator scan_tasks,
+                                   std::shared_ptr<Table>* out) {
   std::vector<std::shared_ptr<RecordBatch>> batches;
 
-  auto it_scantasks = Scan();
-  RETURN_NOT_OK(it_scantasks.Visit([&batches](std::unique_ptr<ScanTask> task) -> Status {
-    auto it = task->Scan();
-    return it.Visit([&batches](std::shared_ptr<RecordBatch> batch) {
-      batches.push_back(batch);
-      return Status::OK();
-    });
-  }));
+  for (auto maybe_scan_task : scan_tasks) {
+    ARROW_ASSIGN_OR_RAISE(auto scan_task, std::move(maybe_scan_task));
+    for (auto maybe_batch : scan_task->Scan()) {
+      ARROW_ASSIGN_OR_RAISE(auto batch, std::move(maybe_batch));
+      batches.emplace_back(batch);
+    }
+  }
 
   return Table::FromRecordBatches(batches, out);
 }
 
+struct TableAggregator {
+  void Append(std::shared_ptr<RecordBatch> batch) {
+    std::lock_guard<std::mutex> lock(m);
+    batches.emplace_back(std::move(batch));
+  }
+
+  Status Finish(std::shared_ptr<Table>* out) {
+    return Table::FromRecordBatches(batches, out);
+  }
+
+  std::mutex m;
+  std::vector<std::shared_ptr<RecordBatch>> batches;
+};
+
+static inline Status ParallelToTable(ScanTaskIterator scan_tasks,
+                                     const ScanContext& context,
+                                     std::shared_ptr<Table>* out) {
+  TableAggregator agg;
+
+  // List ScanTasks without consuming them.
+  std::vector<std::unique_ptr<ScanTask>> tasks;
+  for (auto scan_task : scan_tasks) {
+    RETURN_NOT_OK(scan_task.status());
+    tasks.emplace_back(std::move(scan_task.ValueOrDie()));
```

Comment:
```suggestion
    tasks.emplace_back(std::move(scan_task).ValueOrDie());
```
In this case (since scan_task immediately goes out of scope) the suggestion doesn't correct anything, but it's worth noting that in general `Result` is aware of when it has been moved from (and calling ValueOrDie on a Result which has not been moved will not register that).